compiler/rustc_lint/src/mutable_transmutes.rs

```diff
@@ -0,0 +1,718 @@
+        FieldsShape::Arbitrary { offsets, .. } => offsets.as_slice(),
+        _ => IndexSlice::empty(),
+    };
+    std::iter::zip(field_offsets.iter_enumerated(), field_tys)
+        .map(|((idx, &offset), ty)| (idx, offset, ty))
+}
+
+fn tuple_fields<'tcx>(
+    layout: Layout<'tcx>,
+    field_tys: &'tcx List<Ty<'tcx>>,
+) -> impl Iterator<Item = (FieldIdx, Size, Ty<'tcx>)> + 'tcx {
+    let field_offsets = match &layout.fields {
+        FieldsShape::Arbitrary { offsets, .. } => offsets,
+        _ => bug!("expected FieldsShape::Arbitrary for tuples"),
+    };
+    std::iter::zip(field_offsets.iter_enumerated(), field_tys)
+        .map(|((idx, &offset), ty)| (idx, offset, ty))
+}
+
+#[derive(Debug, Default, Clone)]
+struct FieldsBreadcrumbs(Vec<FieldIdx>);
+
+impl FieldsBreadcrumbs {
+    fn push_with<R>(&mut self, field: FieldIdx, f: impl FnOnce(&mut FieldsBreadcrumbs) -> R) -> R {
+        self.0.push(field);
+        let result = f(self);
+        self.0.pop();
+        result
+    }
+
+    fn translate_for_diagnostics<'tcx>(
+        &self,
+        cx: &LateContext<'tcx>,
+        output: &mut String,
+        mut ty: Ty<'tcx>,
+    ) {
+        for &current_breadcrumb in &self.0 {
+            ty = match ty.kind() {
+                &ty::Adt(adt_def, substs) if adt_def.is_struct() => {
+                    let field = &adt_def.non_enum_variant().fields[current_breadcrumb];
+                    let field_ty =
+                        cx.tcx.normalize_erasing_regions(cx.typing_env(), field.ty(cx.tcx, substs));
+                    output.push_str(".");
+                    output.push_str(field.name.as_str());
+
+                    field_ty
+                }
+
+                &ty::Tuple(tys) => {
+                    let field_ty = tys[current_breadcrumb.as_usize()];
+                    write!(output, ".{}", current_breadcrumb.as_u32()).unwrap();
+
+                    field_ty
+                }
+
+                &ty::Array(element_ty, _) => {
+                    output.push_str("[0]");
+
+                    element_ty
+                }
+
+                _ => bug!("field breadcrumb on an unknown type"),
+            }
+        }
+    }
+
+    // Formats the breadcrumbs as `(*path.to.reference).path.to.unsafecell`;
+    fn format_unsafe_cell_for_diag<'tcx>(
+        cx: &LateContext<'tcx>,
+        reference: &FieldsBreadcrumbs,
+        top_ty: Ty<'tcx>,
+        referent: &FieldsBreadcrumbs,
+        reference_ty: Ty<'tcx>,
+    ) -> TransmuteBreadcrumbs<'tcx> {
+        let mut after_ty = String::new();
+        let before_ty;
+        if !referent.0.is_empty() {
+            before_ty = "(*";
+            reference.translate_for_diagnostics(cx, &mut after_ty, top_ty);
+            after_ty.push_str(")");
+            referent.translate_for_diagnostics(cx, &mut after_ty, reference_ty);
+        } else {
+            before_ty = "*";
+            reference.translate_for_diagnostics(cx, &mut after_ty, top_ty);
```

Comment:
These two lines are shared with the `if` block and can be factored out.

Reply:
In the first `before_ty = "(*"` while in the second it's `"*"`. I find factoring out not worth it.